Mathlib/Tactic/AddTacticDoc.lean

```diff
@@ -0,0 +1,149 @@
+-/
+
+import Std.Util.TermUnsafe
+
+/-!
+# Documentation commands
+
+We generate html documentation from mathlib source code.
+Throughout mathlib, we add documentation entries using `add_tactic_doc` command.
+Then [doc-gen](https://github.com/leanprover-community/doc-gen) calls `getTacticDocEntries`
+to gather all documentation entries, and generates html docs.
+-/
+
+open Lean Elab
+
+/-- The possible categories for `tacticDocEntry.category`. -/
+inductive DocCategory where
+  | tactic   : DocCategory
+  | cmd      : DocCategory
+  | hole_cmd : DocCategory
+  | attr     : DocCategory
+deriving Repr
+
+/-- Prettier name for `docCategory`. -/
+def DocCategory.toString : DocCategory → String
+  | DocCategory.tactic   => "tactic"
+  | DocCategory.cmd      => "command"
+  | DocCategory.hole_cmd => "hole_command"
+  | DocCategory.attr     => "attribute"
+
+/-- The information used to generate a doc entry. -/
+structure TacticDocEntry :=
+  /-- Determines under which name the doc entry is shown in mathlib docs. -/
+  name                     : String
+  /-- Determines under which of predetermined sections (tactics, commands, etc.) this doc entry
+  will lie in mathlib docs. -/
+  category                 : DocCategory
+  /-- Determines the links to which declarations are shown in mathlib docs.
+  Can be used as a source of a docstring (second priority). -/
+  decl_names               : List Lean.Name
+  /-- Free-form tags, e.g. ["simplification", "decision procedure"]. -/
+  tags                     : List String      := []
+  /-- Can be used as a source of a docstring (first priority). -/
+  inherit_description_from : Option Lean.Name := Option.none
+deriving Repr
+
+/-- Turns a `tacticDocEntry` into a JSON representation. -/
+def TacticDocEntry.toJson (tde : TacticDocEntry) (description : String) : Lean.Json :=
+  Lean.Json.mkObj [
+    ("name",       tde.name),
+    ("category",   tde.category.toString),
+    ("decl_names", Lean.Json.arr (List.toArray (tde.decl_names.map (λ x =>
+      Lean.Json.str (toString x)
+    )))),
+    ("tags",        Lean.Json.arr (List.toArray (tde.tags.map Lean.Json.str))),
+    ("description", description)
+  ]
+
+/--
+Serves as a global storage for all `TacticDocEntry`s we have added throughout mathlib codebase.
+When [doc-gen](https://github.com/leanprover-community/doc-gen) generates documentation, it will
+access this storage to find each `tacticDocEntry` along with its greatest-priority docstring.
+-/
+initialize docNamesExtension : SimplePersistentEnvExtension
+  (TacticDocEntry × String) (List (TacticDocEntry × String)) ←
+  registerSimplePersistentEnvExtension {
+    addImportedFn := λ _ => {}
+    addEntryFn    := λ xs x => x :: xs
+  }
+
+/--
+A command used to add documentation for a tactic, command, hole command, or attribute.
+
+Usage: after defining an interactive tactic, command, or attribute,
+add its documentation as follows.
+```lean
+/--
+describe what the command does here
+-/
+add_tactic_doc
```

Comment:
It looks like your version actually requires that the `TacticDocEntry` is split out into a separate def, like this:
```lean
def myTde : tde: TacticDocEntry { name := "display name of the tactic",
  category := cat,
  decl_names := [`dcl_1, `dcl_2],
  tags := ["tag_1", "tag_2"] }
  
  add_tactic_doc myTde
```

Is that right? What is the reason for this change from the mathlib3 version?

Reply:
The change is accidental, it should accept a structure!
Do you have some pointers as to what `tdeStx:______` should be to accept a structure?